### PR TITLE
http: fix issues in CONNECT-UDP forwarding mode.

### PR DESCRIPTION
**Pull request by jeongseokson**
Commit Message:
When Envoy operates as a CONNECT-UDP forwarding proxy, it was resetting the upstream stream because it received HTTP Datagrams before receiving the SETTINGS frame. A new enum has been added in QUICHE to distinguish this case, so I added handling logic for this and made Envoy drop the datagrams instead of resetting the stream.

Also, Envoy was dropping Datagrams because the default maximum packet length for QUIC connections in QUICHE is not large enough for tunneling use cases such as CONNECT-UDP. I added a new QUIC protocol option called `max_packet_length`  to allow users to adjust the maximum packet length for upstream QUIC connections to fix this issue.

Additional Description:
Risk Level: Low, this change is only relevant if CONNECT-UDP is enabled with the forwarding mode.  
Testing: Added more unit tests. 
Docs Changes: Added the `max_packet_length` QUIC protocol option and its explanation.
Release Notes: Added notes about fixing the CONNECT-UDP forwarding mode and adding the new QUIC protocol option.
Platform Specific Features: N/A
[Optional Fixes #Issue]: #34836

### CODE REVIEW

**Comment by zuercher** [comment]
If this is a work-in-progress, please set the status to draft until you're ready for review.

**Comment by jeongseokson** [comment]
Hi Ryan, could you review the update on the HTTP Datagram Handler?
/assign @RyanTheOptimist

Hi Dan, I made some changes in quic_info to set different max packet lengths for QUIC connections if the downstream is CONNECT-UDP. Could you review the changes I made under the QUIC component?
/assign @danzh2010

File: source/common/upstream/cluster_manager_impl.cc
Comment (jeongseokson) [reply source/common/upstream/cluster_manager_impl.cc:2046]: Good point, so the `quic_info_` is always nullptr here in the current code path. A quic info object is assigned in `ProdClusterManagerFactory::allocateConnPool` if `quic_info_` is a nullptr (https://github.com/envoyproxy/envoy/blob/88543c9c37f389ff6d2650bb7aae211563ab0485/source/common/upstream/cluster_manager_impl.cc#L2259). Since this new code is exercised before the `ProdClusterManagerFactory::allocateConnPool`, the current code shouldn't have the dangling pointer issue.

I agree that this is still error-prone though. Another option I was thinking about is just to move this code to `ProdClusterManagerFactory::allocateConnPool` to at least make the instantiation of the quic_info object done in a single place. I was wondering how @danzh2010 would think about the current use of the `quic_info` in the current code, it seems to me that there's room for refactoring.

Comment (jeongseokson) [reply source/common/upstream/cluster_manager_impl.cc:2046]: I think the second question is more important here. My understanding is that the maximum packet length of a QUIC connection is set when it's initialized. This means we should configure all Envoy clusters to create QUIC connections with the larger maximum packet length if a CONNECT-UDP route is configured in the HTTP Connection Manager. Ideally, this configuration would be applied only to the clusters specifically listed under the CONNECT-UDP route, if possible. Does this sound correct, Ryan? I'd also like to hear @DavidSchinazi's thoughts on this.

Comment (jeongseokson) [reply source/common/upstream/cluster_manager_impl.cc:2046]: Another option might be to make this max packet length configurable in Envoy independently. This seems reasonable if we need to adjust this value when connections are created.

Comment (RyanTheOptimist) [reply source/common/upstream/cluster_manager_impl.cc:2046]: Yeah I think this might be a much cleaner way to go. There is already a QuicProtocolOptions field which hsa, for example, `max_concurrent_streams` and `initial_stream_window_size` so putting `max_packet_size` there seems reasonable. https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-quicprotocoloptions

Comment (jeongseokson) [reply source/common/upstream/cluster_manager_impl.cc:2046]: Cool! I think this is the right way to go. I'll send a new commit to the PR soon, but @DavidSchinazi, let me know if you have any other thoughts.

Comment (DavidSchinazi) [reply source/common/upstream/cluster_manager_impl.cc:2046]: Yeah the max packet length needs to be set at QUIC connection setup time, so it can't depend on the method of the request that triggers the connection creation. We could do this based on whether connect-udp is supported at all, but having it be a separate config knob also SGTM

**Review by DavidSchinazi — COMMENTED** [review]
LGTM modulo the open comment thread

File: api/envoy/config/core/v3/protocol.proto
Comment (jeongseokson) [inline api/envoy/config/core/v3/protocol.proto:117]: Thanks for the comments, @RyanTheOptimist and @DavidSchinazi. I will send more commits soon but want to quickly check if we should set a reasonable bound for the max_packet_length option like other options in this config. One range I can think of is 1 <= `max_packet_length` <= 65,535 (max IP packet size), but I'm wondering if we should just not set the bound here. I used `UInt64Value` since the corresponding QUICHE variable and constants (e.g. https://github.com/google/quiche/blob/4249f8025caed1e3d71d04d9cadf42251acb7cac/quiche/quic/core/quic_constants.h#L32) uses `QuicByteCount`, which is `uint64_t` internally. Let me know if you have any thoughts on these.

Comment (DavidSchinazi) [reply api/envoy/config/core/v3/protocol.proto:117]: I don't think Envoy needs to put a cap on this value. Internally, QUICHE will cap it at 1452, but we shouldn't duplicate that constant or logic here

**Review by abeyad — APPROVED** [review]
/lgtm api

File: api/envoy/config/core/v3/protocol.proto
Comment (jeongseokson) [reply api/envoy/config/core/v3/protocol.proto:117]: SGTM. Let me know if you have any other comments on the new code.

File: source/common/upstream/cluster_manager_impl.cc
Comment (jeongseokson) [reply source/common/upstream/cluster_manager_impl.cc]: I added a new `max_packet_length` QUIC protocol option. We don't update the `quic_info_` pointer anymore in the new code so the first question from Ryan and my answer to that is no longer relevant. Let me know any of you have other comments.

**Comment by jeongseokson** [comment]
@RyanTheOptimist, @DavidSchinazi Let me know if you have more comments, @abeyad I made your approval obsolete so will appreciate it if you can LGTM again, thank you.

**Review by DavidSchinazi — APPROVED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
@abeyad, over to you, I think!

**Review by abeyad — APPROVED** [review]
/lgtm api

**Comment by jeongseokson** [comment]
It passed the CI, but I'm not sure how to remove the waiting label. Can anyone with write access merge this PR? Thank you.

**Pull request merged**
